### PR TITLE
Add support for conditionally linking dependencies based on platform

### PR DESCRIPTION
**Pull request by daltonclaybrook**
This PR makes it possible to conditionally link dependencies based on a specified list of platforms. It addresses an issue raised in #951 that the "platform filter" parameter in Xcode is currently not granular enough to prevent a framework from linking against tvOS and watchOS targets.

This is currently implemented as a new field on `Dependency` called `conditionalPlatforms`. In all honestly, I would prefer that the existing `platform` setting be renamed to something like `platformFilter` to disambiguate from this new setting, but I'm open to suggestions.

(cc: @raptorxcz, @yonaskolb, @jlott1)

### CODE REVIEW

**Comment by daltonclaybrook** [comment]
Spitballing here: What if we changed the other field to be called `platformFilter` and this one simply `platforms`. Possible changes to docs:

>- [ ] **platformFilter**: **String** - This field is specific to Mac Catalyst. It corresponds to the "Platforms" dropdown in the Frameworks & Libraries section of Target settings in Xcode. Available options are: **iOS**, **macOS** and **all**. Defaults is **all**
>- [ ] **platforms**: **[Platform](#platform)** - List of platforms this dependency should apply to. Defaults to all applicable platforms.

File: Sources/ProjectSpec/Dependency.swift [hunk 1]
```diff
@@ -137,6 +144,7 @@ extension Dependency: JSONEncodable {
             "embed": embed,
             "codeSign": codeSign,
             "link": link,
+            "conditionalPlatforms": conditionalPlatforms.map(Array.init)?.map(\.rawValue)
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Dependency.swift:147]: Can we add a sort to make this deterministic. And I also don't think the Array init is required

**Review by yonaskolb — COMMENTED** [review]
Thanks @daltonclaybrook. I agree with the rename of platformFilter, so please proceed. It can be marked as a breaking change in the changelog.

**Comment by daltonclaybrook** [comment]
@yonaskolb Just pushed up those changes. How do you feel about the name of the new field (currently "`conditionalPlatforms`")? Should this be changed as well?

**Comment by daltonclaybrook** [comment]
@yonaskolb Please let me know what the next steps are on this. I'd really like to get this landed since my team needs it for an upcoming project.

**Comment by yonaskolb** [comment]
> @yonaskolb Just pushed up those changes. How do you feel about the name of the new field (currently "`conditionalPlatforms`")? Should this be changed as well?

Yes, let's rename this to `platforms`. Sorry for the delay. Will approve after your change

**Review by yonaskolb — APPROVED** [review]
Great, thanks @daltonclaybrook!

**Pull request merged**